### ISSUE TITLE
[ListView] prepend without full re-render

### ISSUE THREAD

**Issue by akabab** [facebook/react-native#2219]
How can the list be aware of positionnal changes without trashing all the layout, like adding a new element up front or deleting an existing element. Currently, the row check is based on it's Array index, when shifted even if the exact same element is only a row below, it will be considered as a completly different value and so redrawn. This is disastrous for performance :/

[StackOverflow for code](http://stackoverflow.com/questions/31774399/react-native-listview-how-to-push-new-rows-from-top)

Am I missing something?

**Comment by jdmunro** [comment]
We've encountered a performance issue with similar circumstances when prepending to a list with over 100 rows. I noticed you're modifying the value for key by prepending some text. I don't think this is required, in fact, I think React may even be expecting the key to be the rowId exactly. Our performance issue was fixed by using the rowId as the key, whereas previously we were generating our own unique IDs.

**Comment by akabab** [comment]
anyone to confirm the key fix cause it's not fixing in my case, I can't find the real utility of keys.

**Comment by sghiassy** [comment]
This might help improve your performance: https://github.com/sghiassy/react-native-sglistview

**Comment by maicaiyao** [comment]
We've met the same problem when we want to animate the prepending list.

**Comment by hayeah** [comment]
TLDR: make sure that the keys you use are stable. Array indices wouldn't do.

@maicaiyao @akabab when you have a list of children, you should use React's `key` attribute. Each key should uniquely identify the row. Suppose you have two rows:

```
<div>
  <div key="a">...</div>
  <div key="b">...</div>
</div>
```

If you swap them around, like so:

```
<div>
  <div key="b">...</div>
  <div key="a">...</div>
</div>
```

React would recognize that they are the same rows as before, and do a reordering instead of rerender.

The problem you hare having is that you are using array index as the key (probably). In which case if you delete or insert an item, the `key` attribute may change for ALL items. This is what causes React to re-render everything.

**Comment by brentvatne** [comment]
@hayeah has some good tips, also check out: https://github.com/exponentjs/react-native-invertible-scroll-view

**Comment by ide** [comment]
Yep always use a unique non-index based keys for this. See the React docs & code for more info about the reconciliation algo.

**Comment by akabab** [comment]
@hayeah oh yeah thx!
